vars/xsDeploy.groovy

```diff
@@ -0,0 +1,329 @@
+              */
+            .withMandatoryProperty('mtaPath', null, {action == Action.NONE})
+            .withMandatoryProperty('deploymentId',
+                'No deployment id provided, neither via parameters nor via common pipeline environment. Was there a deployment before?',
+                 {action in [Action.RESUME, Action.ABORT, Action.RETRY]})
+            .use()
+
+        utils.pushToSWA([
+            step: STEP_NAME,
+        ], config)
+
+        if(performLogin) {
+            login(script, config)
+        }
+
+        def failures = []
+
+        if(action in [Action.RESUME, Action.ABORT, Action.RETRY]) {
+
+            complete(script, mode, action, config, failures)
+
+        } else {
+
+            deploy(script, mode, config, failures)
+        }
+
+        if (performLogout || failures) {
+            logout(script, config, failures)     
+
+        } else {
+            echo "Skipping logout in order to be able to resume or abort later."
+        }
+
+        if(failures) {
+            error "Failed command(s): ${failures}. Check earlier log for details."
+        }
+    }
+}
+
+void login(Script script, Map config) {
+
+    withCredentials([usernamePassword(
+        credentialsId: config.credentialsId,
+        passwordVariable: 'password',
+        usernameVariable: 'username'
+    )]) {
+
+        def returnCode = executeXSCommand([script: script].plus(config.docker),
+        [
+            "xs login -a ${config.apiUrl} -u ${username} -p ${BashUtils.quoteAndEscape(password)} -o ${config.org} -s ${config.space} ${config.loginOpts}",
+            'RC=$?',
+            "[ \$RC == 0 ]  && cp \"\${HOME}/${config.xsSessionFile}\" .",
+            'exit $RC'
+        ])
+
+        if(returnCode != 0)
+            error "xs login failed."
+    }
+
+    boolean existsXsSessionFileAfterLogin = fileExists file: config.xsSessionFile
+    if(! existsXsSessionFileAfterLogin)
+        error "Session file ${config.xsSessionFile} not found in current working directory after login."
+}
+
+void deploy(Script script, DeployMode mode, Map config, def failures) {
+            
+    def deploymentLog
+
+    try {
+        lock(getLockIdentifier(config)) {
+            deploymentLog = executeXSCommand([script: script].plus(config.docker),
+            [
+                "cp ${config.xsSessionFile} \${HOME}",
+                "xs ${mode.toString()} '${config.mtaPath}' -f ${config.deployOpts}"
+            ], true)
+        }
+
+        echo "Deploy log: ${deploymentLog}"
+
+    } catch(AbortException e) {
+        echo "deployment failed. Message: ${e.getMessage()}, Log: ${deploymentLog}}"
+        failures << "xs ${mode.toString()}"
+    }
+
+    if(mode == DeployMode.BG_DEPLOY) {
+
+        if(! failures.isEmpty()) {
+
+            echo "Retrieval of deploymentId skipped since prior deployment was not successfull."
+
+        } else {
+
+            for (def logLine : deploymentLog.readLines()) {
```

Comment:
Actually I hate log file parsing. But there is no other known way how the deployment id is exposed (e.g. as environment variable). Hence we use log file parsing as the last resort.